lib/foreman/http_proxy/rest_client_extension.rb

```diff
@@ -8,6 +8,23 @@ def proxy_uri
         end
         super
       end
+
+      def net_http_object(hostname, port)
+        p_uri = proxy_uri
+
+        if p_uri.nil?
+          # no proxy set
+          Net::HTTP.new(hostname, port)
+        elsif !p_uri
+          # proxy explicitly set to none
+          Net::HTTP.new(hostname, port, nil, nil, nil, nil)
+        else
+          proxy_pass = CGI.unescape(p_uri.password) if p_uri.password
+          proxy_user = CGI.unescape(p_uri.user) if p_uri.user
```

Comment:
Since we monkey patch `RestClient`, this can break every call to it if they don't escape. I wonder if we should use a special object like `CgiEscapedString` and only then unescape. Thoughts?

Reply:
The username and password is passed through an URI object, which performs validations on them and raises an exception if they do not pass. This reduces the issue to someone using escape sequences literally in their username/password, where unescaping them would yield something different.

Can we even do that sensibly? It would have to be a subclass of String to be accepted by URI, but then if it got stored in the db and loaded again, it would be a plain String. We'd have to start marking the db entries similarly to how we handle encrypted strings. Is it worth it?

Reply:
Probably not, but I am worried about monkey patching another library. However, I think the way RestClient works you can't really subclass it to EscapedRestClient or similar, right? What if another third party library also uses RestClient?

And I do realize the chance is small because someone would need to have something that's CGI escaped as a password but you just know that someone at some point will have that.

Reply:
As far as I understand them, refinements have a limited scope and we need this to be available anywhere a user might use restclient. Or am I missing something?

Reply:
That's the idea, to limit the scope of RestClient changes. I would think that since we change the way the parameters are passed here, we can also add a factory method to create modified `RestClient::Request` objects here. The only drawback I can see is if the developer wants to use the direct syntax: `RestClient::Request.execute(`, but IMHO it's a small price to pay for getting everything encapsulated.

Reply:
I still don't see how refinements would help us without us changing the code everywhere where we use restclient,

Reply:
That's true. We won't be able to get away without changing our own code.I was referring to the opposite problem (as far as I understand Ewoud's comment), where we are influencing RestClient instances the we don't own at all.
I can't think about a solution that is both transparent and does not influence external libraries at this point. In my opinion keeping the change scoped only to our project takes precedence on transparency of the change.

Reply:
Done in https://github.com/theforeman/foreman/pull/8412/commits/119273b889745b4b9372146f7441b92e2ac481d8 , but it may also require followup patches to foreman_ansible, katello and possibly other plugins.